### ISSUE TITLE
Constant in computed key breaks literal string types in values

### ISSUE THREAD

**Issue by pimterry** [Microsoft/TypeScript#21116]
**TypeScript Version:**  2.6.2

**Code**

```ts
type X = {
    mykey: { type: 'mytype' }
}

const key = 'mykey';
let x: X = {
    [key]: { type: 'mytype' },
};
```

**Expected behavior:**

This code should compile.

**Actual behavior:**

It fails to compile, with:

```ts
test.ts(6,5): error TS2322: Type '{ [key]: { type: string; }; }' is not assignable to type 'X'.
  Types of property 'mykey' are incompatible.
    Type '{ type: string; }' is not assignable to type '{ type: "mytype"; }'.
      Types of property 'type' are incompatible.
        Type 'string' is not assignable to type '"mytype"'.
```

If `[key]` is replaced with `mykey` or `['mykey']`, either of which should be equivalent, then it _does_ compile.

It seems that using a computed key from a variable, even when its exact type is known for sure, causes the compiler to broaden the type of the value from literal `'mytype'` to `string`.

**Comment by andy-ms** [comment]
I don't see any error in `typescript@next`. May have already been fixed since you listed `2.6.2` above.

**Comment by andy-ms** [comment]
We generally close an issue as soon as it's fixed in `typescript@next`.
